cpp/ql/src/semmle/code/cpp/controlflow/Guards.qll

```diff
@@ -8,303 +8,387 @@ import semmle.code.cpp.controlflow.Dominance
+  cached
+  predicate comparesEq(Expr left, Expr right, int k, boolean areEqual, boolean testIsTrue) {
+    compares_eq(this, left, right, k, areEqual, testIsTrue)
+  }
+
+  /**
+   * Holds if (determined by this guard) `left == right + k` must be `areEqual` in `block`.
+   *        If `areEqual = false` then this implies `left != right + k`.
+   */
+  cached
+  predicate ensuresEq(Expr left, Expr right, int k, BasicBlock block, boolean areEqual) {
+    exists(boolean testIsTrue |
+      compares_eq(this, left, right, k, areEqual, testIsTrue) and this.controls(block, testIsTrue)
+    )
+  }
+
+  /**
+   * Holds if this condition controls `block`, meaning that `block` is only
+   * entered if the value of this condition is `testIsTrue`. This helper
+   * predicate does not necessarily hold for binary logical operations like
+   * `&&` and `||`. See the detailed explanation on predicate `controls`.
+   */
+  private predicate controlsBlock(BasicBlock controlled, boolean testIsTrue) {
+    exists(BasicBlock thisblock | thisblock.contains(this) |
+      exists(BasicBlock succ |
+        testIsTrue = true and succ = this.getATrueSuccessor()
+        or
+        testIsTrue = false and succ = this.getAFalseSuccessor()
+      |
+        bbDominates(succ, controlled) and
+        forall(BasicBlock pred | pred.getASuccessor() = succ |
+          pred = thisblock or bbDominates(succ, pred) or not reachable(pred)
         )
-    }
-
-    /**
-     * Holds if this condition controls `block`, meaning that `block` is only
-     * entered if the value of this condition is `testIsTrue`. This helper
-     * predicate does not necessarily hold for binary logical operations like
-     * `&&` and `||`. See the detailed explanation on predicate `controls`.
-     */
-    private predicate controlsBlock(BasicBlock controlled, boolean testIsTrue) {
-        exists(BasicBlock thisblock
-        | thisblock.contains(this)
-        | exists(BasicBlock succ
-          | testIsTrue = true and succ = this.getATrueSuccessor() or
-            testIsTrue = false and succ = this.getAFalseSuccessor()
-          | bbDominates(succ, controlled) and
-            forall(BasicBlock pred
-            | pred.getASuccessor() = succ
-            | pred = thisblock or bbDominates(succ, pred) or not reachable(pred))))
-    }
+      )
+    )
+  }
 }
 
 private predicate is_condition(Expr guard) {
-    guard.isCondition() or
-    is_condition(guard.(BinaryLogicalOperation).getAnOperand()) or
-    exists(NotExpr cond | is_condition(cond) and cond.getOperand() = guard)
+  guard.isCondition()
+  or
+  is_condition(guard.(BinaryLogicalOperation).getAnOperand())
+  or
+  exists(NotExpr cond | is_condition(cond) and cond.getOperand() = guard)
 }
 
-
- /* Simplify conditions in the source to the canonical form l op r + k.
+/*
+ * Simplification of equality expressions:
+ * Simplify conditions in the source to the canonical form l op r + k.
```

Comment:
I don't understand what's happened here.

Reply:
I was confused as to why the two lines have swapped order.  It looks like it was a manual change, and probably does make more sense than the initial version.

Reply:
I swapped the lines since that seemed to make more sense and for consistency with the "Simplification of inequality expressions" comment further down.